### PR TITLE
ENH: add FIR and IIR gammatone filters to scipy.signal

### PR DESCRIPTION
**Pull request by skresearcher**
#### What does this implement/fix?
Adds finite impulse response and infinite impulse response gammatone filters to the scipy.signal module.

### CODE REVIEW

File: scipy/signal/filter_design.py [hunk 1]
```diff
@@ -4843,6 +4845,152 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    order = 1
+    return ((hz / EarQ) ** order + minBW ** order) ** (1 / order)
+
+
+def iirgammatone(freq, fs=16000, output='ba'):
+    """
+    IIR gammatone filter design.
+
+    This function computes the coefficients of an 8th order IIR
+    digital filter, which models a 4th order gammatone filter.
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4864]:
```suggestion
    digital filter [1]_, which models a 4th order gammatone filter.
```

File: scipy/signal/filter_design.py [hunk 2]
```diff
@@ -4843,6 +4845,152 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    See Also
+    --------
+    firgammatone
+    iirfilter
+    iirdesign
+
+    References
+    ----------
+    Slaney, Malcolm, "An Efficient Implementation of the
+    Patterson–Holdsworth Auditory Filter Bank", pp.34-39.
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4907]:
```suggestion
    .. [1] Slaney, Malcolm, "An Efficient Implementation of the
           Patterson–Holdsworth Auditory Filter Bank", pp.34-39.
```

File: scipy/signal/filter_design.py [hunk 3]
```diff
@@ -4843,6 +4845,152 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    Utility for converting from frequency (Hz) to the
+    Equivalent Rectangular Bandwith (ERB) scale
+    """
+    EarQ = 9.26449
+    minBW = 24.7
+    order = 1
+    return ((hz / EarQ) ** order + minBW ** order) ** (1 / order)
+
+
+def iirgammatone(freq, fs=16000, output='ba'):
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4859]: - I'm not sure it's worth having the `output` here. It's computed as `tf` form, it's easy enough (one-liner/wrapped) for people to convert at the end.
- There should either be no default `fs`, or it should be `1.` or `2 * np.pi` as in other places in the code. My vote here would probably be for no default

File: scipy/signal/fir_filter_design.py [hunk 4]
```diff
@@ -1260,3 +1262,103 @@ def minimum_phase(h, method='homomorphic', n_fft=None):
...
+
+    # Check for invalid input
+    if not isinstance(freq, int) and not isinstance(freq, float):
+        raise ValueError("The cuttoff frequency must be an integer or float.")
+    if not isinstance(order, int):
+        raise ValueError("The order of the filter must be an integer.")
+    if not isinstance(fs, int):
+        raise ValueError("The sampling frequency must be an integer.")
+    if not isinstance(numtaps, int):
+        raise ValueError("numtaps must be an integer.")
```
Comment (larsoner) [inline scipy/signal/fir_filter_design.py:1340]: This is not how we check for int-ness, for example this will not work for `np.int` and related types. Usually we just do:
```suggestion
    numtaps = operator.index(numtaps)
```

File: scipy/signal/fir_filter_design.py [hunk 5]
```diff
@@ -1260,3 +1262,103 @@ def minimum_phase(h, method='homomorphic', n_fft=None):
...
+    if not isinstance(scale, bool):
+        raise ValueError("scale must be a boolean.")
```
Comment (larsoner) [inline scipy/signal/fir_filter_design.py:1342]: No need to be so strict here, I would just do `scale = bool(scale)` or really not bother (it will happen implicitly later)

File: scipy/signal/fir_filter_design.py [hunk 6]
```diff
@@ -1260,3 +1262,103 @@ def minimum_phase(h, method='homomorphic', n_fft=None):
...
+    >>> freq = 440
+    >>> signal.firgammatone(freq, numtaps=16)
+    array([ 0.00000000e+00,  2.68586728e-07,  1.97702315e-06,  5.90239274e-06,
+        1.18022128e-05,  1.82536512e-05,  2.26825900e-05,  2.15748723e-05,
+        1.08452242e-05, -1.36746652e-05, -5.56777430e-05, -1.17807768e-04,
+       -2.01126526e-04, -3.04675651e-04, -4.25179658e-04, -5.56926298e-04])
+    """
+
+    # Check for invalid input
+    if not isinstance(freq, int) and not isinstance(freq, float):
```
Comment (larsoner) [inline scipy/signal/fir_filter_design.py:1333]:
```suggestion
    freq = float(freq)
```
This will error out nicely enough if something inappropriate is given

**Review by larsoner — COMMENTED** [review]
What about a unified design like:
```
def gammatone(freq, fs, order=int | None, numtaps=240 | None, scale=True)
```
- if `order is None` then it uses IIR, if int it uses FIR
- if `numtaps is None` you use `int(round(fs * 0.015))` (this is the optimal ratio, right?)

It unifies the interface, and always just returns `b, a` for the filter (IIR or FIR depending on the `order` argument).

Also, why is there a `scale` argument for `FIR` but not `IIR`? It seems like they should either both have the `scale` argument or neither should.

**Comment by skresearcher** [comment]
@larsoner thanks for reviewing this PR.

> What about a unified design like:
> 
> ```
> def gammatone(freq, fs, order=int | None, numtaps=240 | None, scale=True)
> ```
> 
>     * if `order is None` then it uses IIR, if int it uses FIR
> 
>     * if `numtaps is None` you use `int(round(fs * 0.015))` (this is the optimal ratio, right?)
> 
> 
> It unifies the interface, and always just returns `b, a` for the filter (IIR or FIR depending on the `order` argument).

The reason I split the `FIR` and `IIR` gammatone functions is because SciPy has separate files for `FIR` and `IIR` filters: therefore, it seemed prudent to follow the existing layout. Currently, the `butter()` and `firwin()` functions are in separate files as well. Let me know if you prefer to combine the `FIR` and `IIR` Gammatone filters. I can also combine `butter()` and `firwin()` if necessary to keep it consistent. 

> Also, why is there a `scale` argument for `FIR` but not `IIR`? It seems like they should either both have the `scale` argument or neither should.

I'll remove the `scale` parameter from the `FIR` filter.

File: scipy/signal/filter_design.py [hunk 7]
```diff
@@ -4843,6 +4845,152 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    """
+    IIR gammatone filter design.
+
+    This function computes the coefficients of an 8th order IIR
+    digital filter, which models a 4th order gammatone filter.
+
+    Parameters
+    ----------
+    freq : float
+        Center frequency of the filter in Hz.
```
Comment (WarrenWeckesser) [inline scipy/signal/filter_design.py:4869]: Drive-by comment: Don't specify explicit units (Hz) here.  This is the frequency, in whatever units are determined by `fs`.

**Comment by larsoner** [comment]
> The reason I split the FIR and IIR gammatone functions is because SciPy has separate files for FIR and IIR filters: therefore, it seemed prudent to follow the existing layout. 

I think this is mostly because there are typically FIR filters (firwin, firwin2, etc.) and IIR filters (butter, etc.) and not any that can be either. Also keep in mind from a user perspective these both end up in the `scipy.signal` namespace, that their actual implementation is in separate `.py` files is merely a developer-related detail. It makes more sense to me to have a single function where you specify the output type somehow (IIR or FIR) than to have different names.

**Comment by skresearcher** [comment]
I've merged the FIR and IIR gammatone filters into one function, updated the unit tests, and incorporated @larsoner's suggestions on the sampling rate default value and error-checking.

File: scipy/signal/filter_design.py [hunk 8]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+
+    See Also
+    --------
+    firwin
+    iirfilter
+
+    References
+    ----------
+    .. [1] Slaney, Malcolm, "An Efficient Implementation of the
+        Patterson–Holdsworth Auditory Filter Bank", pp.3-8, 34-39.
```
Comment (ilayn) [inline scipy/signal/filter_design.py:4906]: Ref missing the report type or source name.

File: scipy/signal/filter_design.py [hunk 9]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+            if fs <= 1000:
+                numtaps = 15
+            else:
+                numtaps = int(fs * 0.015)
+        else:
+            numtaps = operator.index(numtaps)
+
+        # Check for invalid order
+        if order <= 0 or order > 24:
+            raise ValueError("Invalid order: order must be > 0 and <= 24.")
```
Comment (ilayn) [inline scipy/signal/filter_design.py:4972]: This should be mentioned to the user in the docstring

File: scipy/signal/filter_design.py [hunk 10]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    if ftype == 'fir':
+        # Set order and numtaps if not passed
+        if order is None:
+            order = 4
+        else:
+            order = operator.index(order)
+
+        if numtaps is None:
+            if fs <= 1000:
+                numtaps = 15
```
Comment (ilayn) [inline scipy/signal/filter_design.py:4964]: This should be mentioned in the docstring. No hidden defaults...by default.

File: scipy/signal/filter_design.py [hunk 11]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    Gammatone filter design.
+
+    This function computes the coefficients of an FIR or IIR gammatone
+    digital filter [1].
+
+    Parameters
+    ----------
+    freq : float
+        Center frequency of the filter (expressed in the same units
+        as `fs`).
```
Comment (ilayn) [inline scipy/signal/filter_design.py:4870]: What's the unit when fs is not given since it is optional?

File: scipy/signal/filter_design.py [hunk 12]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    if fs is None:
+        fs = 2
+    else:
+        fs = operator.index(fs)
+
+    # Check for invalid cutoff frequency or filter type
+    filter_types = ['fir', 'iir']
+    if freq <= 0 or freq >= fs / 2:
+        raise ValueError("Invalid cutoff frequency: frequency must be "
+                         "> 0 and < fs/2.")
```
Comment (ilayn) [inline scipy/signal/filter_design.py:4950]: You can tell the valid values to the user instead of text errors, for example something like

```python
raise ValueError("The frequency must be between 0 and {}"
                 " (nyquist), but given {}.".format(fs/2, freq))
```

**Review by ilayn — CHANGES_REQUESTED** [review]
I don't have time to go into the math part unfortunately, but there a few places where the results would underflow quickly. I'll check hopefully later.

File: scipy/signal/filter_design.py [hunk 13]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
     return b, a
 
 
+def _hz_to_erb(hz):
```
[the remark this answers is not in the document]
Comment (skresearcher) [reply scipy/signal/filter_design.py:4848]: This formula is used in the MATLAB function for IIR gammatone filters described in the paper. It is a rewritten version of the original formula described in Glasberg and Moore's “Derivation of auditory filter shapes from notched-noise data.” (p.2)

File: scipy/signal/filter_design.py (hunk 10 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py:4964]: This was already included in the docstring.

File: scipy/signal/filter_design.py (hunk 11 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py:4870]: The default value for fs is 2, like firwin() and iirfilter(). Added it in the docstring.

File: scipy/signal/filter_design.py (hunk 12 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py:4950]: I based this error message from firwin(), but I changed it now. The invalid cutoff frequency error message is not consistent between all the filter functions.

File: scipy/signal/filter_design.py [hunk 14]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    digital filter [1].
+
+    Parameters
+    ----------
+    freq : float
+        Center frequency of the filter (expressed in the same units
+        as `fs`).
+    ftype : {'fir', 'iir'}
+        The type of filter the function generates.
+    order : int, optional
```
[the remark this answers is not in the document]
Comment (skresearcher) [reply scipy/signal/filter_design.py:4873]: I've added more detail about the order of the FIR and IIR filter in the docstring.

File: scipy/signal/filter_design.py [hunk 15]
```diff
@@ -4843,6 +4845,203 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+
+    Returns
+    -------
+    b, a : ndarray, ndarray
+        Numerator (``b``) and denominator (``a``) polynomials of the filter.
+
+    Raises
+    ------
+    ValueError
+        If `freq` is less than or equal to 0
```
[the remark this answers is not in the document]
Comment (skresearcher) [reply scipy/signal/filter_design.py:4892]: Fixed.

File: scipy/signal/filter_design.py (hunk 8 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py:4906]: Fixed.

File: scipy/signal/filter_design.py (hunk 9 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py:4972]: Fixed.

File: scipy/signal/filter_design.py (hunk 11 above)
Comment (ilayn) [reply scipy/signal/filter_design.py:4870]: But what is the unit, 2 hz.?

File: scipy/signal/filter_design.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ilayn) [reply scipy/signal/filter_design.py:4848]: That's possible but that would be bad coding on the author's side. This is a naive implementation of the p-norm which is prone to underflows. Picked the first two results from a search 

https://rehanguha.github.io/articles/2019-03/overflow-p-norm
https://timvieira.github.io/blog/post/2014/11/10/numerically-stable-p-norms/

File: scipy/signal/filter_design.py (hunk 11 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py:4870]: Please refer to an earlier comment on this. Also, in firwin() it does not specify the unit as Hz for fs.
https://github.com/scipy/scipy/pull/12200#discussion_r430610263

File: scipy/signal/filter_design.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (skresearcher) [reply scipy/signal/filter_design.py:4848]: I've made a short script comparing the different hz2erb conversion formulas with detailed comments. You can check it at the link here:

https://github.com/skresearcher/audiotools/blob/master/hz2erb.py

To deal with the overflow issue, I've simplified the math and removed the `order` exponent from the conversion, as it was just set to 1 and didn't affect the calculation.

**Comment by skresearcher** [comment]
The TravisCI and Windows errors are unrelated to this PR: I tried to re-run the checks by pushing empty commits, but they keep failing.

File: scipy/signal/filter_design.py (hunk 10 above)
Comment (larsoner) [reply scipy/signal/filter_design.py:4967]: This can be simplified to (seems more readable and less branching):
```
if numtaps is None:
    numtaps = max(int(fs * 0.015), 15)
numtaps = operator.index(numtaps)
```

File: scipy/signal/filter_design.py [hunk 16]
```diff
@@ -4843,6 +4845,206 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    minBW = 24.7
+    return hz / EarQ + minBW
+
+
+def gammatone(freq, ftype, order=None, numtaps=None, fs=None):
+    """
+    Gammatone filter design.
+
+    This function computes the coefficients of an FIR or IIR gammatone
+    digital filter [1].
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4864]:
```suggestion
    digital filter [1]_.
```
As is, this will not render properly, see CircleCI output (please check to make sure you're otherwise happy with it):

https://19633-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.gammatone.html#scipy.signal.gammatone

File: scipy/signal/filter_design.py [hunk 17]
```diff
@@ -4843,6 +4845,206 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+        if numtaps is None:
+            if fs <= 1000:
+                numtaps = 15
+            else:
+                numtaps = int(fs * 0.015)
+        else:
+            numtaps = operator.index(numtaps)
+
+        # Check for invalid order
+        if order <= 0 or order > 24:
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4974]: Perhaps also more readable
```suggestion
        if not 0 < order <= 24:
```

File: scipy/signal/filter_design.py [hunk 18]
```diff
@@ -4843,6 +4845,206 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+    >>> plt.show()
+    """
+    # Converts freq to float
+    freq = float(freq)
+
+    # Set sampling rate if not passed
+    if fs is None:
+        fs = 2
+    else:
+        fs = operator.index(fs)
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4947]: I don't think we should require `fs` to be an int, so this should just be:
```
fs = float(fs)
```
and you can unident it if you want. I would, it's not going to affect performance and the less we need to think about what happens in each branch the better. By unindenting, the code becomes "if None, set a default; next line, ensure fs is a float"...

File: scipy/signal/filter_design.py [hunk 19]
```diff
@@ -4843,6 +4845,206 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+
+    # Check for invalid cutoff frequency or filter type
+    filter_types = ['fir', 'iir']
+    if freq <= 0 or freq >= fs / 2:
+        raise ValueError("The frequency must be between 0 and {}"
+                         " (nyquist), but given {}.".format(fs / 2, freq))
+    if ftype.lower() not in filter_types:
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4954]: you check `ftype.lower()` here but then do `ftype == 'fir'` below. You either need `ftype.lower()` in the conditional below or you need to remove it here

File: scipy/signal/tests/test_filter_design.py [hunk 20]
```diff
@@ -3734,3 +3734,73 @@ def test_w_or_N_types(self):
...
+    def test_invalid_input(self):
+        # Cutoff frequency is <= 0 or >= fs / 2.
+        fs = 16000
+        assert_raises(ValueError, gammatone, -fs, 'iir', fs=fs)
+        assert_raises(ValueError, gammatone, 0, 'fir', fs=fs)
+        assert_raises(ValueError, gammatone, fs / 2, 'iir', fs=fs)
+        assert_raises(ValueError, gammatone, fs, 'fir', fs=fs)
+
+        # Filter type is not fir or iir
+        assert_raises(ValueError, gammatone, 440, ftype='fie', fs=16000)
```
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3750]: It's clearer to do something like:
```
with pytest.raises(ValueError, match='ftype must be'):
    gammatone(440, ftype='fie', fs=16000)
```
Most of our tests don't have this because historically we couldn't make use of `pytest.raises`, but now we can, so newer code can benefit from it.

File: scipy/signal/filter_design.py (hunk 16 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed.

File: scipy/signal/filter_design.py (hunk 17 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed. Also fixed `freq` between 0 and `fs/2`.

File: scipy/signal/filter_design.py (hunk 18 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed.

File: scipy/signal/filter_design.py (hunk 19 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed. I convert `ftype` to lowercase before checking all of the conditions.

File: scipy/signal/tests/test_filter_design.py (hunk 20 above)
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: Fixed. I switched the remaining `assert_raises` tests to `pytest.raises` and also switched the `assert_approx_equal` tests to use `pytest.approx`.

File: scipy/signal/filter_design.py (hunk 1 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed.

File: scipy/signal/filter_design.py [hunk 21]
```diff
@@ -4843,6 +4845,201 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
...
+        if numtaps is None:
+            numtaps = max(int(fs * 0.015), 15)
+        numtaps = operator.index(numtaps)
+
+        # Check for invalid order
+        if not 0 < order <= 24:
+            raise ValueError("Invalid order: order must be > 0 and <= 24.")
+
+        # Gammatone impulse response settings
+        t = np.linspace(0, numtaps / fs, numtaps)
```
Comment (larsoner) [inline scipy/signal/filter_design.py:4973]: These will not be spaced by `1. / fs`, e.g.:
```
>>> np.linspace(0, 10 / 1000., 10)
array([0.        , 0.00111111, 0.00222222, 0.00333333, 0.00444444,
       0.00555556, 0.00666667, 0.00777778, 0.00888889, 0.01      ])
```
To get that you need:
```
t = np.linspace(0, (numtaps - 1) / fs, numtaps)
```
or
```
t = np.linspace(0, numtaps / fs, numtaps, endpoint=False)
```
or (my preference)
```
t = np.arange(numtaps) / fs
```
Also, it's not great that the tests did not catch this discrepancy -- it suggests that they are either incomplete, or tolerances are too high.

File: scipy/signal/tests/test_filter_design.py [hunk 22]
```diff
@@ -3734,3 +3734,76 @@ def test_w_or_N_types(self):
...
+        b2 = [1.31494461367464e-06, -5.03391196645395e-06,
+              7.00649426000897e-06, -4.18951968419854e-06,
+              9.02614910412011e-07]
+        a2 = [1.0, -7.65646235454218,
+              25.7584699322366, -49.7319214483238,
+              60.2667361289181, -46.9399590980486,
+              22.9474798808461, -6.43799381299034,
+              0.793651554625368]
+        b == pytest.approx(b2)
+        a == pytest.approx(a2)
```
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3809]: In SciPy it's more standard to use `assert_allclose` because it gives precise control over absolute and relative tolerances. So probably better to use it here rather than introduce `pytest.approx`

Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py]: Fixed. I kept `pytest.approx` on the frequency response unit test because it's just comparing one number and it's faster than `assert_approx_equal`.

File: scipy/signal/filter_design.py (hunk 21 above)
Comment (skresearcher) [reply scipy/signal/filter_design.py]: Fixed. I added another unit test comparing the FIR filter output with the paper's Mathematica implementation.

File: scipy/signal/tests/test_filter_design.py [hunk 23]
```diff
@@ -3787,14 +3787,28 @@ def test_frequency_response(self):
...
+        assert_array_equal(sorted(z), sorted(z.conj()))
+        assert_array_equal(sorted(p), sorted(p.conj()))
+        assert_equal(k, np.real(k))
 
         assert_(issubclass(b.dtype.type, np.floating))
         assert_(issubclass(a.dtype.type, np.floating))
 
-    # Verify IIR filter coefficients with MATLAB's answers
+    # Verify FIR filter coefficients with the paper's
+    # Mathematica implementation
```
Comment (larsoner) [inline scipy/signal/tests/test_filter_design.py:3798]: If it was wrong the first time you computed it in MATLAB, it's worth adding a note here about the difference you found so that the next person who cross-checks isn't confused by the difference they find

**Review by larsoner — COMMENTED** [review]
Other than one more code comment that would be helpful, LGTM.

@toddrjen can you look?

File: scipy/signal/tests/test_filter_design.py (hunk 23 above)
Comment (skresearcher) [reply scipy/signal/tests/test_filter_design.py:3798]: The paper does not have a MATLAB implementation of FIR gammatone filters, only IIR gammatone filters. So I cross-checked my answers with the paper's Mathematica formulas instead. I have included the page references in the main function. Do you want me to include the references in the unit test comments as well?

Comment (larsoner) [reply scipy/signal/tests/test_filter_design.py:3798]: Ahh okay, I misunderstood. I think this is fine as is!

**Comment by larsoner** [comment]
@ilayn you are marked as requesting changes as well, can you look again?

File: scipy/signal/filter_design.py [hunk 24]
```diff
@@ -4863,6 +4865,201 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
     return b, a
 
 
+def _hz_to_erb(hz):
```
Comment (toddrjen) [inline scipy/signal/filter_design.py:4868]: Is there a reason this has to be private?  It is a useful function in its own right.  I know it is simple, but if it is going to be in there, I think it might as well be available for use.

**Review by larsoner — APPROVED** [review]
Failures are unrelated and things LGTM. Let's go ahead and merge early next week unless someone else has comments (feel free to ping me to remind me to do it)

**Comment by skresearcher** [comment]
@larsoner can you please merge this PR in?

**Comment by larsoner** [comment]
Thanks @skresearcher !
